### DISCUSSION TITLE
Array of objects showing as undefined and thus, not rendering.

### DISCUSSION THREAD

**Discussion by nashvegasnate** [nss-evening-cohort-14/evening-client-side#314] [Capstones!] [answered]
Name: Nathan Keith

Issue: When I try and navigate to my page of user song lists (titled "singleList.js"), nothing renders to the DOM. Devtools shows nothing. Url shows "localhost:3000/lists/undefined". Each user has their own set of songs and lists, and I have an intermediary table set up in firebase to handle the many-to-many relationship (since many songs can belong to many lists and many lists can have many songs). I have a promise.all (listsWithSongs) that is intended to map through the array of lists belonging to a specific user and then filter to show the songs on a particular list when the user clicks on "view songs" button. However, I can't get the song cards to show up in the container.

Tickets referenced:
#301 
#306 

What I have Googled: 
https://stackoverflow.com/questions/32157286/rendering-react-components-from-array-of-objects
https://forum.freecodecamp.org/t/use-array-map-to-dynamically-render-elements/198828
https://www.pluralsight.com/guides/manipulating-arrays-and-objects-in-state-with-react

What I have tried:
- implemented the useState hook with an empty default value as its parameters
- utilizing the useEffect hook and passing it the listsWithSongs function

GitHub links to files:
https://github.com/nashvegasnate/SetList/blob/nk-readSongLists/src/helpers/data/ListSongsData.js
https://github.com/nashvegasnate/SetList/blob/nk-readSongLists/src/views/Lists.js
https://github.com/nashvegasnate/SetList/blob/nk-readSongLists/src/helpers/Routes.js

**Comment by drteresavasquez** [comment]
Do you have a link to the file that is setting the URL? Also your routes don't have anything to handle /lists/:whatever so you need to handle that once you figure out why it's loading undefined. That means you're pushing an undefined value into your Params.

**Reply by drteresavasquez** [reply]
@nashvegasnate in this line where is the id coming from on listSongs.id? You're setting listSongs as an empty array in your state so that will be undefined. https://github.com/nashvegasnate/SetList/blob/33b42c48c9cc1aaa8d0678d804ae91ae81c9f513/src/views/SingleList.js#L13

**Reply by nashvegasnate** [reply]
"id" is the identifier in the join table (listSongs) that links songs with associated lists. Does that make sense? I'm sure I'm using it as incorrectly as possible; I'm just struggling at the moment. And were you really up at 1:30am responding to my ticket? That is dedication! I appreciate it immensely :)

**Reply by drteresavasquez** [reply]
Ha! Sleepless nights 😊. So the id is not being passed to the url properly. That's the first issue. The second is that you don't have a route to handle it when you are passing it right. What I would do to trouble shoot is create the route and then copy an id from Firebase and paste it into your url and see what happens. If nothing happens you know all of this to be the case. 

We can take a look tomorrow in class. Also if you haven't already add it to your flow chart so it makes sense how you're setting the url param and grabbing it.

**Comment by drteresavasquez** [comment]
As discussed, let us know when your single API calls are working. Once those are done, we can focus on the complex data function and coach you through that process. You are almost there!

**Reply by nashvegasnate** [reply]
All my individual API calls are giving me back the requested data, as verified by Postman. I will definitely need some coaching!

**Reply by drteresavasquez** [reply]
Did you try to get it working?

**Reply by nashvegasnate** [reply]
Yes, but still nothing.

**Comment by nashvegasnate** [comment, marked as answer]
My original complex data function was calling all functions prior to .then, which meant that they weren't all ready to return the data when the function ran, thus resulting in an empty array. The beginning of my function looked similar to this:
```const listsWithSongs = (firebaseKey, uid) => new Promise((resolve, reject) => {
Promise.all([getSongs(uid), getLists(uid), getListSongs(firebaseKey)])
.then(([songs, lists, listsWithSongs]) => {
.....
```
After talking through it with Aja, it was obvious that the Promise.all needed to be written after the songs and listSongs had been declared. Also, it was not necessary to include getLists since what we needed were the songs on the user's specific list of songs (listSongs) rather than all the lists. The songs on the lists were identified by the listId and not firebaseKey, so that was also substituted. The final, working promise.all looked like this:
```
const listsWithSongs = (uid, listId) => new Promise((resolve, reject) => {
  const songs = getSongs(uid);
  const listSongs = getListSongs(listId);
  Promise.all([songs, listSongs])
    .then(([songsResponse, listSongsResponse]) => {
      const listRelationshipsArray = listSongsResponse.filter((sl) => sl.listId === listId);

      const songInfoArray = listRelationshipsArray.map((listRelationship) => songsResponse.find((song) => song.firebaseKey === listRelationship.songId));
      resolve(songInfoArray);
    }).catch((error) => reject(error));
});
```
The song cards are now populating on the singleList view! Although, the song cards are not rendering all the information in their complete form, but that may be the next discussion ticket...
